src/services/events/index.ts

```diff
@@ -0,0 +1,55 @@
+import { HeseyaEvents, EventCallbackFunction, HeseyaEventToPayloadMap } from './utils/types'
+
+export interface HeseyaEventListenerService {
+  emit: <Key extends keyof HeseyaEvents>(
+    event: Key,
+    ...payloads: HeseyaEventToPayloadMap[Key] extends undefined
+      ? [undefined?]
+      : [HeseyaEventToPayloadMap[Key]]
+  ) => void
+  on: <Key extends keyof HeseyaEvents>(
+    event: Key,
+    cb: EventCallbackFunction<HeseyaEventToPayloadMap[Key]>,
+  ) => void
+  unsubscribe: <Key extends keyof HeseyaEvents>(
+    event: Key,
+    cb: EventCallbackFunction<HeseyaEventToPayloadMap[Key]>,
+  ) => void
+}
+
+export const createHeseyaEventListenerService = (): HeseyaEventListenerService => {
+  const map: HeseyaEvents = {
```

Comment:
Może zamiast tego typu użyć po prostu `Record<HeseyaEventType, EventCallbackFunction<unknown>[]>`?

Wewnętrzna implementacja nie musi znać typów dopóki zewnętrzne interfejsy są otypowane poprawnie. Nie będzie wtedy trzeba dawać tyle assercji co masz teraz + jak już to na typ `unknown` który jest bezpieczniejszy niż `any`

Jeśli jeszcze zamienisz wszystkie `keyof HeseyaEvents` na `HeseyaEventType` to wtedy można też całkiem pozbyć się typu `HeseyaEvents`, który już teraz jest mocno redundantny bo duplikuje payloady i eventy